### PR TITLE
Implemented Safe Enum Migration

### PR DESCRIPTION
**Pull request by Antonwy**
## Problem Statement
You want to add a new enum type to a table column, inside a migration. The existing way to add a type e.g. `canteenHeadCount` is, to alter the table in the following way:

```sql
ALTER TABLE crontab MODIFY COLUMN type enum('news', 'mensa', 'chat', 'kino', 'roomfinder', 'ticketsale', 'alarm', 'fileDownload','dishNameDownload','averageRatingComputation', 'canteenHeadCount');
```

That could cause "race conditions" between branches.

To avoid this I've created the migration functions `SafeEnumMigrate` and `SafeEnumRollback`.

### `SafeEnumMigrate`
```go
func SafeEnumMigrate(tx *gorm.DB, table interface{}, column string, additionalTypes ...string) error
```
Takes the migration transaction `tx`, the table model `table` which includes the enum column, the column name `column` and one or more additional types `additionalTypes`.

- `SafeEnumMigrate` then parses the current column enum type and adds the `additionalTypes` to the back of the enum type list

### `SafeEnumRollback `
```go
func SafeEnumRollback(tx *gorm.DB, table interface{}, column string, rollbackTypes ...string) error
```

Works like `SafeEnumMigrate` just takes `rollbackTypes` which are then removed from the enum type list.

## Testing
Unit tests can be found in `safeEnumMigrate_test.go`

### CODE REVIEW

File: server/backend/migration/safeEnumMigrate.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	return list
+}
+
+func alterEnumColumn(tx *gorm.DB, table interface{}, column string, types []string) error {
+	enum := BuildEnum(types)
+
+	stmt := &gorm.Statement{DB: tx}
+	err := stmt.Parse(&table)
+	if err != nil {
+		return errors.New("could not parse table")
```
Comment (COM8) [inline server/backend/migration/safeEnumMigrate.go:53]:
```suggestion
		return errors.New("could not parse enum table")
```

File: server/backend/migration/safeEnumMigrate.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		"ALTER TABLE %s MODIFY %s %s;",
+		tableName,
+		column,
+		enum,
+	)
+
+	tx = tx.Exec(rawQuery)
+
+	if tx.Error != nil {
+		return errors.New("could not alter table")
```
Comment (COM8) [inline server/backend/migration/safeEnumMigrate.go:67]:
```suggestion
		return errors.New("could not alter enum table")
```

**Review by COM8 — CHANGES_REQUESTED** [review]
Thanks for looking into this!

Looks really promising.

* In golang tests can access functions that are not exported. So I don't see a reason, why for example `BuildEnum` should be upper case.
* Right now you just added the functionality, but you did not migrate the code to use this feature. Is this intended?

**Review by joschahenningsen — APPROVED** [review]
Exported methods could use some comments, looks good otherwise :+1:

**Pull request merged**